hstream-store/HStream/Store.hs

```diff
@@ -1,9 +1,163 @@
+import qualified Data.Map.Strict         as Map
+import           Data.Word               (Word32)
+import           GHC.Generics            (Generic)
+import           Z.Data.CBytes           (CBytes)
+import qualified Z.Data.JSON             as JSON
+import           Z.Data.Vector           (Bytes)
+
 import           HStream.Store.Exception
 import           HStream.Store.Logger
-import           HStream.Store.Stream
+import qualified HStream.Store.Stream    as S
+
+-------------------------------------------------------------------------------
+
+data ProducerRecord = ProducerRecord
+  { dataInTopic     :: S.Topic
+  , dataInKey       :: Maybe CBytes
+  , dataInValue     :: Bytes
+  , dataInTimestamp :: Int64
+  } deriving (Show, Generic, JSON.JSON)
+
+newtype ProducerConfig = ProducerConfig { producerConfigUri :: CBytes }
+  deriving (Show, Generic, JSON.JSON)
+
+newtype Producer = Producer S.StreamClient
+
+mkProducer :: ProducerConfig -> IO Producer
+mkProducer config = do
+  client <- S.newStreamClient (producerConfigUri config)
+  return $ Producer client
+
+sendMessage :: Producer -> ProducerRecord -> IO ()
+sendMessage (Producer client) record@ProducerRecord{..} = do
+  topicID <- S.getTopicIDByName client dataInTopic
+  void $ S.appendSync client topicID (JSON.encode record) Nothing
+
+-- FIXME: performance improvements
+sendMessages :: Producer -> [ProducerRecord] -> IO ()
+sendMessages producer xs = forM_ xs $ sendMessage producer
+
+-------------------------------------------------------------------------------
+
+data ConsumerRecord = ConsumerRecord
+  { dataOutTopic     :: S.Topic
+  , dataOutOffset    :: S.SequenceNum
+  , dataOutKey       :: Maybe CBytes
+  , dataOutValue     :: Bytes
+  , dataOutTimestamp :: Int64
+  } deriving (Show, Generic, JSON.JSON)
+
+dataRecordToConsumerRecord :: S.DataRecord -> ConsumerRecord
+dataRecordToConsumerRecord S.DataRecord{..} = do
+  case JSON.decode' recordPayload of
+    Left _err -> error "JSON decode error!"
+    Right ProducerRecord{..} ->
+      ConsumerRecord { dataOutTopic     = dataInTopic
+                     , dataOutOffset    = recordLSN
+                     , dataOutKey       = dataInKey
+                     , dataOutValue     = dataInValue
+                     , dataOutTimestamp = dataInTimestamp
+                     }
+
+data ConsumerConfig = ConsumerConfig
+  { consumerConfigUri         :: CBytes
+  , consumerName              :: CBytes
+    -- ^ Unique identifier of one consumer
+  , consumerBufferSize        :: Int64
+  -- ^ specify the read buffer size for this client, fallback
+  -- to the value in settings if it is -1
+  , consumerCheckpointUri     :: CBytes
+  , consumerCheckpointRetries :: Word32
+  } deriving (Show, Generic, JSON.JSON)
+
+data Consumer = Consumer
+  { unConsumer      :: S.StreamSyncCheckpointedReader
+  , _consumerTopics :: Map S.Topic S.TopicID
+  }
+
+mkConsumer :: ConsumerConfig -> [S.Topic] -> IO Consumer
```

Comment:
It looks like the consumer is strongly tied to some topics, so what should a consumer do if it wants to change topics of their subscription?